documentation/contributing/syncing_backporting/index.markdown

```diff
@@ -0,0 +1,205 @@
+---
+author: henningkayser
+comments: true
+date: 2021-11-05 07:00:00+01:00
+layout: contribution-guide
+slug: syncing
+title: Syncing and Backporting
+redirect_from: "/documentation/contributing/syncing_backporting.html"
+---
+
+# Syncing and Backporting
+
+In order to make bug-fixes and features available across repos and distros, we need to migrate them from time to time.
+Syncing refers to the process of preparing a merge commit from one branch to another that keeps the commit history intact.
+This allows us to better keep track of all the changes that need to get synced, but also helps with resolving issues later on since the original history of the source branch is available in the correct order for later review.
+We usually only do this from the MoveIt 1 master branch to the `main` branch of MoveIt 2, but other repos could also use the same strategy.
+
+Backporting is the act of migrating isolated pull reqests or commits from one branch to another using a cherry-pick.
+The typical use case is that we backport bug-fixes from the MoveIt 2 `main` branch to stable distro branches or to MoveIt 1.
+
+Below are step-by-step guides that should help maintainers to perform a proper sync or backport.
+
+## Syncing from MoveIt 1 to MoveIt 2
+
+Let's assume `origin` is the git remote for MoveIt 2 and `moveit1` the one for MoveIt 1, like below (using SSH works just the same, of course):
+
+
+**origin:** https://github.com/ros-planning/moveit2.git
+
+**moveit1:** https://github.com/ros-planning/moveit.git
+
+
+
+### 1. Create a new sync branch
+
+Fetch the latest MoveIt 2 `main` branch and use it for creating a new sync branch `pr-sync`:
+
+```
+git fetch origin
+git checkout -b pr-sync origin/main
+```
+
+
+### 2. Identify the MoveIt 1 commit that was used for the last sync
+
+The `merge-base` command gives you the hash of the common base of the two branches.
+
+```
+git fetch moveit1
+LAST_SYNC_COMMIT=$(git merge-base origin/main moveit1/master)
+echo $LAST_SYNC_COMMIT
+```
+
+
+You should be able to verify the merge commit hash of the last sync by searching for it in the commit log of the `main` branch using the command below.
+
+```
+git log --oneline | grep $LAST_SYNC_COMMIT
+```
+
+It should print a merge commit message similar to `"Merge https://github.com/ros-planning/moveit/commit/<commit_hash>"`.
+If you don't find a matching commit message the last maintainer didn't follow this guide.
+In that case you can try searching for the merge commit by only filtering for the first couple digits of the hash.
+
+
+### 3. Pick a new sync commit
+
+You can now identify the new commit in MoveIt 1 that should be used for your MoveIt 2 sync.
+The comment below gives you all commits that have not yet been synced:
+
+```
+git log --oneline moveit1/master $(LAST_SYNC_COMMIT)..HEAD
+```
+
```

Comment:
Another way to visualize all commits that have not been synced

```suggestion
git config --global alias.hist "log --pretty=format:'%C(yellow)[%ad]%C(reset) %C(green)[%h]%C(reset) | %C(red)%s %C(bold red){{%an}}%C(reset) %C(blue)%d%C(reset)' --graph --date=short"
git hist main..moveit1/master
```

Reply:
I agree that aliases like this can make git outputs prettier or even more useful. To me, it would be a little out of scope to add this here or even a bit distracting. Much better would be a reference like: "Hint: you can also use the `git hist` alias instead of "git log" for some nicer output" along with a link to a source or another "Git Best Practices" section.
I'll leave this change out for now, but feel free to contribute something like this.

Reply:
It's actually a real improvement in terms of readability. :+1:
We don't have a better place to put it, so sprinkling neat tool improvements in such instructions is still nice.